src/zcap.rs

```diff
@@ -0,0 +1,473 @@
+    // embedded proofs such as LD-PROOF
+    //   https://w3c-ccg.github.io/ld-proofs/
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proof: Option<Proof>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub property_set: Option<Map<String, Value>>,
+}
+
+impl<A, C> Delegation<A, C>
+where
+    A: Serialize + Send + Sync + Clone,
+    C: Serialize + Send + Sync + Clone,
+{
+    pub async fn verify(
+        &self,
+        _options: Option<LinkedDataProofOptions>,
+        resolver: &dyn DIDResolver,
+    ) -> VerificationResult {
+        match &self.proof {
+            None => VerificationResult::error("No applicable proof"),
+            Some(proof) => {
+                let mut result = proof.verify(self, resolver).await;
+                if result.errors.is_empty() {
+                    result.checks.push(Check::Proof);
+                }
+                result
+            }
+        }
+    }
+
+    // https://w3c-ccg.github.io/ld-proofs/
+    pub async fn generate_proof(
+        &self,
+        jwk: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<Proof, Error> {
+        LinkedDataProofs::sign(self, options, jwk).await
+    }
+
+    /// Prepare to generate a linked data proof. Returns the signing input for the caller to sign
+    /// and then pass to [`ProofPreparation::complete`] to complete the proof.
+    pub async fn prepare_proof(
+        &self,
+        public_key: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<ProofPreparation, Error> {
+        LinkedDataProofs::prepare(self, options, public_key).await
+    }
+
+    pub fn set_proof(self, proof: Proof) -> Self {
+        Self {
+            proof: Some(proof),
+            ..self
+        }
+    }
+}
+
+#[async_trait]
+impl<A, C> LinkedDataDocument for Delegation<A, C>
+where
+    A: Serialize + Send + Sync + Clone,
+    C: Serialize + Send + Sync + Clone,
+{
+    fn get_contexts(&self) -> Result<Option<String>, Error> {
+        Ok(Some(serde_json::to_string(&self.context)?))
+    }
+
+    async fn to_dataset_for_signing(
+        &self,
+        parent: Option<&(dyn LinkedDataDocument + Sync)>,
+    ) -> Result<DataSet, Error> {
+        let mut copy = self.clone();
+        copy.proof = None;
+        let json = serde_json::to_string(&copy)?;
+        let more_contexts = match parent {
+            Some(parent) => parent.get_contexts()?,
+            None => None,
+        };
+        let mut loader = StaticLoader;
+        json_to_dataset(&json, more_contexts.as_ref(), false, None, &mut loader).await
+    }
+}
+
+// limited initial definition of a ZCAP Invocation, generic over Action
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Invocation<A> {
+    #[serde(rename = "@context")]
+    pub context: Contexts,
+    pub id: URI,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub capability_action: Option<A>,
```

Comment:
The spec doesn't mention `capabilityAction`, but uses `action`: https://w3c-ccg.github.io/zcap-ld/#actions
But the security v2 context has `capabilityAction` and not `action`, so I understand that would be why this is used. :+1:
Related issue: https://github.com/w3c-ccg/zcap-ld/issues/17

Reply:
Indeed, I figured it was better to use the close existing term than to pull in another context. As menttioned in that issue, would it be better to use the type field? e.g. `@type: ["CapabilityInvocation", "Read"]`?

Reply:
Using `@type` seems reasonable to me; I don't know which is better. If other implementations are doing different things, we might have to support all three.

Reply:
`@type` also seems reasonable to me, while DB's [zcap-ld seems to use capabilityAction](https://github.com/digitalbazaar/zcapld/blob/main/lib/CapabilityInvocation.js#L26). The spec only loosely suggests having a field called `action`:
> Targets are free to choose their own mechanisms for directing behavior, but MAY support the action property on invocations as one common behavioral direction technique.
We could be extensible by:
- adding a flattened domain-specific field of type `T` to both docs, as a type param e.g. `Invocation<A, T>`
- using some kind of trait which must be implemented on one/both docs for `Invocation::verify` to work, which devs can implement to represent their action validation logic?
e.g. for `capabilityAction`, the type `T` is a struct `{ capabilityAction: String }`, and the trait is implemented for docs using that `T` as their extra data